### PR TITLE
Move containerd process into docker cgroup for versions >= v1.11

### PR DESCRIPTION
**Pull request by timstclair**
Addresses https://github.com/kubernetes/kubernetes/issues/23397#issuecomment-209583923

/cc @vishh @kubernetes/sig-node

### CODE REVIEW

File: pkg/kubelet/cm/container_manager_linux.go [hunk 1]
```diff
@@ -472,42 +480,58 @@ func getPidsForProcess(name, pidFile string) ([]int, error) {
...
 // Ensures that the Docker daemon is in the desired container.
-func ensureDockerInContainer(cadvisor cadvisor.Interface, oomScoreAdj int, manager *fs.Manager) error {
-	pids, err := getPidsForProcess(dockerProcessName, dockerPidFile)
-	if err != nil {
-		return err
+func ensureDockerInContainer(dockerVersion semver.Version, oomScoreAdj int, manager *fs.Manager) error {
+	type process struct{ name, file string }
+	dockerProcs := []process{{dockerProcessName, dockerPidFile}}
+	if dockerVersion.GTE(containerdVersion) {
+		dockerProcs = append(dockerProcs, process{containerdProcessName, ""})
```
Comment (vishh) [inline pkg/kubelet/cm/container_manager_linux.go]: Is there a pid-file for containerd?

File: pkg/kubelet/cm/container_manager_linux.go [hunk 2]
```diff
@@ -472,42 +480,58 @@ func getPidsForProcess(name, pidFile string) ([]int, error) {
...
-		// Also apply oom-score-adj to processes
-		oomAdjuster := oom.NewOOMAdjuster()
-		if err := oomAdjuster.ApplyOOMScoreAdj(pid, oomScoreAdj); err != nil {
-			errs = append(errs, fmt.Errorf("failed to apply oom score %d to PID %d", oomScoreAdj, pid))
+func ensureProcessInContainer(pid int, oomScoreAdj int, manager *fs.Manager) error {
+	if runningInHost, err := isProcessRunningInHost(pid); err != nil {
+		// Err on the side of caution. Avoid moving the docker daemon unless we are able to identify its context.
+		return err
+	} else if !runningInHost {
+		// Docker daemon is running inside a container. Don't touch that.
```
Comment (vishh) [inline pkg/kubelet/cm/container_manager_linux.go]: nit: Update comment.

**Comment by vishh** [comment]
LGTM. Thanks for the cleanup @timstclair !

File: pkg/kubelet/cm/container_manager_linux.go (hunk 1 above)
Comment (timstclair) [inline pkg/kubelet/cm/container_manager_linux.go]: I decided to try one more search, and found it! `/run/docker/libcontainerd/docker-containerd.pid` I can't find any documentation of that though...

File: pkg/kubelet/cm/container_manager_linux.go (hunk 2 above)
Comment (timstclair) [inline pkg/kubelet/cm/container_manager_linux.go]: done.

**Comment by timstclair** [comment]
Fixed. Thanks!

File: pkg/kubelet/cm/container_manager_linux.go (hunk 1 above)
Comment (vishh) [inline pkg/kubelet/cm/container_manager_linux.go]: I'm not surprised about the lack of documentation. Let's add it.

File: pkg/kubelet/cm/container_manager_linux.go [hunk 3]
```diff
@@ -54,8 +55,15 @@ const (
 	// The minimum memory limit allocated to docker container: 150Mi
 	MinDockerMemoryLimit = 150 * 1024 * 1024
 
-	dockerProcessName = "docker"
-	dockerPidFile     = "/var/run/docker.pid"
+	dockerProcessName     = "docker"
+	dockerPidFile         = "/var/run/docker.pid"
+	containerdProcessName = "docker-containerd"
+	containerdPidFile     = "/run/docker/libcontainerd/docker-containerd.pid"
```
Comment (vishh) [inline pkg/kubelet/cm/container_manager_linux.go]: Was this `/var/run` or `/run`? Just checking..

**Comment by vishh** [comment]
@timstclair If there is a PR that is switching docker versions to v1.11, then I'd suggest marking this PR as a pre-req for that PR.

File: pkg/kubelet/cm/container_manager_linux.go (hunk 3 above)
Comment (timstclair) [inline pkg/kubelet/cm/container_manager_linux.go]: `/run` is right (it has more docker information than `/var/run`)

File: pkg/kubelet/cm/container_manager_linux.go (hunk 1 above)
Comment (timstclair) [inline pkg/kubelet/cm/container_manager_linux.go]: Oops, sorry! Forgot to update this to the new constant.

**Comment by timstclair** [comment]
Thanks, done.

**Comment by timstclair** [comment]
Reapplying LGTM after trivial fix (addressing @vishh's comment)

**Comment by timstclair** [comment]
@k8s-bot node e2e test this, issue #26385

File: pkg/kubelet/cm/container_manager_linux.go (hunk 3 above)
Comment (philips) [inline pkg/kubelet/cm/container_manager_linux.go]: Is using this pid file documented and stable? We should probably ask the Docker team to do that if not.

Comment (timstclair) [inline pkg/kubelet/cm/container_manager_linux.go]: I can't find any documentation around this path, but we do have a fallback (`pidof`) if the file isn't found. I'll follow up with a docker issue to check the stability.

**Comment by dchen1107** [comment]
According to the perfdash, cpu usage of docker drops 50% after introducing docker 1.11.1, but that is not the truth since this pr is not merged, and part of resource overhead is not accounting to docker cgroup. :-( Bumping the priority to p0 so that we can have a better resource overhead monitoring for 1.3.

**Comment by timstclair** [comment]
@k8s-bot e2e test this, issue https://github.com/kubernetes/kubernetes/issues/26324

**Pull request merged**